TestBackfillChannels Failing Test

```
> --- FAIL: TestBackfillChannels_FetchesPerChannelSinceSyncedAt (0.24s)
>     reconnect_backfill_test.go:152: missing upserted message C1/150: sql: no rows in result set
> FAIL
> FAIL      github.com/gammons/slk/cmd/slk  2.793s
> FAIL
```

Trying to build v0.7.7 from source this test consistently fails. Bisecting it looks like the test started failing for me in v0.7.6.

Failing in the middle of that test feels weird but I haven't had enough time to look at the code to see why it might be failing on my system.

Building with go 1.26.2.

Thanks!

--- a/internal/cache/db.go
+++ b/internal/cache/db.go
@@ -3,6 +3,7 @@ package cache
 import (
 	"database/sql"
 	"fmt"
+	"strings"
 
 	_ "modernc.org/sqlite"
 )
@@ -11,22 +12,40 @@ type DB struct {
 	conn *sql.DB
 }
 
+// dsnPragmas are appended to every DSN passed to New(). They are
+// applied per-connection by modernc.org/sqlite as the pool opens new
+// connections, which is the only reliable way to set per-connection
+// pragmas (a one-off conn.Exec only sets the pragma on the single
+// connection that ran it). See issue #9: without busy_timeout, two
+// goroutines that write at the same time fail the second with
+// SQLITE_BUSY immediately instead of waiting, which the reconnect
+// backfill silently swallowed.
+//
+//   - busy_timeout(5000): writers wait up to 5s for a competing
+//     writer to finish before returning SQLITE_BUSY. Five seconds
+//     comfortably covers the bursty fan-out in runChannelPhase.
+//   - journal_mode(WAL): concurrent readers don't block the writer.
+//     WAL persists in the file header once set, but applying it
+//     per-connection is harmless and keeps it visible in the DSN
+//     alongside busy_timeout.
+//   - foreign_keys(ON): mirrors the previous one-shot PRAGMA exec.
+const dsnPragmas = "_pragma=busy_timeout(5000)&_pragma=journal_mode(WAL)&_pragma=foreign_keys(ON)"
+
+// appendPragmas returns dsn with the per-connection pragmas spliced
+// into its query string. Handles plain paths, ":memory:", and
+// already-URI-formatted DSNs (file:..., or path?key=value).
+func appendPragmas(dsn string) string {
+	sep := "?"
+	if strings.Contains(dsn, "?") {
+		sep = "&"
+	}
+	return dsn + sep + dsnPragmas
+}
+
 func New(dsn string) (*DB, error) {
-	conn, err := sql.Open("sqlite", dsn)
+	conn, err := sql.Open("sqlite", appendPragmas(dsn))
 	if err != nil {
 		return nil, fmt.Errorf("opening database: %w", err)
-	}
-
-	// Enable WAL mode for better concurrent read performance
-	if _, err := conn.Exec("PRAGMA journal_mode=WAL"); err != nil {
-		conn.Close()
-		return nil, fmt.Errorf("setting WAL mode: %w", err)
-	}
-
-	// Enable foreign keys
-	if _, err := conn.Exec("PRAGMA foreign_keys=ON"); err != nil {
-		conn.Close()
-		return nil, fmt.Errorf("enabling foreign keys: %w", err)
 	}
 
 	db := &DB{conn: conn}

--- a/internal/cache/db_test.go
+++ b/internal/cache/db_test.go
@@ -1,6 +1,7 @@
 package cache
 
 import (
+	"context"
 	"database/sql"
 	"path/filepath"
 	"testing"
@@ -147,6 +148,52 @@ func TestMigrateAddsChannelsSyncedAtColumn(t *testing.T) {
 	}
 	if !found {
 		t.Error("channels table missing synced_at column")
+	}
+}
+
+// TestNew_SetsBusyTimeoutOnAllPoolConnections is a regression test
+// for issue #9. Without a busy_timeout pragma on every connection in
+// the pool, two goroutines in WAL mode that try to write at the same
+// time will fail the second writer with SQLITE_BUSY immediately
+// instead of waiting. The reconnect backfill (cmd/slk/reconnect_backfill.go)
+// fans out N goroutines across the shared *sql.DB and silently
+// dropped messages on systems where the lock window was long enough
+// to collide.
+//
+// We force the pool to open multiple connections and assert that
+// each one has a non-zero busy_timeout. PRAGMA busy_timeout is
+// per-connection in sqlite, so the only way to ensure every pooled
+// connection has it is to set it in the DSN (so it runs as part of
+// the per-connection init), not via a one-off conn.Exec.
+func TestNew_SetsBusyTimeoutOnAllPoolConnections(t *testing.T) {
+	db, err := New(filepath.Join(t.TempDir(), "busy.db"))
+	if err != nil {
+		t.Fatalf("New: %v", err)
+	}
+	defer db.Close()
+
+	ctx := context.Background()
+	// Hold N concurrent conns so the pool actually opens N distinct
+	// underlying sqlite connections. If we just queried PRAGMA on
+	// db.conn N times, the pool would hand back the same connection.
+	const N = 4
+	conns := make([]*sql.Conn, 0, N)
+	for i := 0; i < N; i++ {
+		c, err := db.conn.Conn(ctx)
+		if err != nil {
+			t.Fatalf("Conn[%d]: %v", i, err)
+		}
+		conns = append(conns, c)
+	}
+	for i, c := range conns {
+		var bt int
+		if err := c.QueryRowContext(ctx, "PRAGMA busy_timeout").Scan(&bt); err != nil {
+			t.Fatalf("conn %d PRAGMA busy_timeout: %v", i, err)
+		}
+		if bt < 1000 {
+			t.Errorf("conn %d busy_timeout = %d ms, want >= 1000 (writers must wait, not return SQLITE_BUSY immediately)", i, bt)
+		}
+		c.Close()
 	}
 }
 

